### ISSUE TITLE
Add week-of-year to semantic skeletons

### ISSUE THREAD

**Issue by sffc** [unicode-org/icu4x#5643]
Week-of-year should be added to semantic skeletons, both in CLDR and in ICU4X.

Previous issue: https://github.com/unicode-org/icu4x/issues/488

Week-of-year requires plurals support. Plurals are mostly supported in the data model already, so this should not require a breaking change.

The following fixture tests were removed in #5640 and should be added back when ready:

```json
    {
        "description": "Date time example that includes a week-of for a locale with plural variants as well as one with non default WeekData",
        "input": {
            "locale": "en",
            "value": "2016-04-17T08:25:07.000+05:00",
            "options": {
                "components": {
                    "year": "numeric",
                    "week": "numeric-week-of-year",
                    "hour": "two-digit",
                    "minute": "two-digit",
                    "second": "two-digit",
                    "time_zone_name": "offset"
                }
            }
        },
        "output": {
            "values": {
                "en": "week 17 of 2016, 08:25:07 GMT+05:00",
                "en-AU": "week 16 of 2016, 08:25:07 GMT+05:00",
                "en-GB": "week 15 of 2016, 08:25:07 GMT+05:00",
                "fil": "linggo 17 ng 2016, 08:25:07 GMT+05:00",
                "en-ZA": "week 17 of 2016, 08:25:07 GMT+05:00"
            }
        }
    }
```

**Comment by sffc** [comment]
Previously, week-of-year was handled internally in the datetime formatter. However, in the semantic skeleton world and with flexible input types, I would rather move the week number calculations out into the input type. For example, I would rather have something such as

```javascript
pub struct WeekDate<C: WeekCalendar> {
    inner: C::WeekDateInner
}

struct ArithmeticWeek {
    pub(crate) year: i32,
    pub(crate) week: u8,
    pub(crate) weekday: Weekday, // 1-7
}

pub struct GregorianWeekDateInner(pub(crate) ArithmeticWeek);

impl WeekCalendar for Gregorian {
    type WeekDateInner = GregorianWeekDateInner;
}
```

The functions for converting between `Date` and `WeekDate` involve data and locale preferences (i.e. `WeekCalculator`).

(this is not a 2.0-issue, but CC @Manishearth @robertbastian @zbraniecki)

**Comment by robertbastian** [comment]
I would like to at least come up with a design for this before 2.0. Combining the date and the locale from the formatter is going to complicate the API, and semver needs to be thought through now.

**Comment by sffc** [comment]
Yeah fair point.

There are two designs already: the old one (where WeekCalculator is loaded in the formatter) and the one suggested in https://github.com/unicode-org/icu4x/issues/5643#issuecomment-2408973458 (which is a major shift, encoding it in the calendar type).

A middle ground would be to basically treat this similar to weekday. We don't calculate the weekday inside the formatter, even though it should in theory be derivable from the year, month, and day. Likewise, we shouldn't necessarily calculate the week number inside the formatter. The more I think about this, the more I realize that calculating the week number in the formatter is putting logic in the wrong place: it is at its core a calendrical calculation, but the formatter should only be passing through data unchanged, just looking up display names.

I would love it if week numbering were by calendar, but unfortunately it has to include the locale. This means that `Date<A>` can't calculate the week number.

One option is that we make `LocalizedDate<A>`, which implements `GetField<WeekOfYearInfo>`. A downside of this direction is that the locale could be out of sync between the `LocalizedDate` and the `DateTimeFormatter`.

**Comment by sffc** [comment]
As far as the semantic skeleton API goes:

My thinking for a while has been that it should just be a new field set. `Year + Week`, a new Calendar Period Field Set, and `Year + Week + Weekday`, a new Date Field Set.

Question to people who use week calendars: does anyone ever say "Wednesday, Week 30 of 2024" as a stand-in for "July 24, 2024", and should semantic skeletons support locales that prefer year-week-weekday over year-month-day? Or is it sufficient to make the developer opt-in to week calendars? @eggrobin

**Comment by robertbastian** [comment]
Day-of-week is currently modeled by an `IsoWeekday`, but that's not sufficient for patterns `e..ee` and `c..cc`. These require a locale-sensitive days-since-start-of-week number, so we probably need to introduce a `FormattableDayOfWeek`, which is returned only by `LocalizedDate`.

That said, week data is *tiny*, it's less than 500 bytes (lookup table + data), and it's not data that changes on the regular. It would be a shame to overcomplicate our API for data size or custom-data reasons.

**Comment by sffc** [comment]
- @sffc I see this as a calendrical calculation
- @robertbastian DayOfYearInfo has all the info you need
- @Manishearth I think it is a display concern
- @robertbastian - seconded
- @sffc How can we make it calendar-dependent, not just locale-dependent
- @robertbastian Add a field to DayOfYearInfo. Oops, it should be non_exhaustive
- @sffc Can we add this post-2.0?
- @robertbastian It is better if we start loading WeekCalculator data structs in datetime now, then not loading them later will not be breaking, but the other way would be
- @sffc I agree that this is the easiest way forward, even though I still feel it would be more elegant to solve this in icu_calendar, but I think the membrane placement with DayOfYearInfo is not too horrible. OK.

LGTM: @sffc @robertbastian

**Comment by sffc** [comment]
2.0 work is to load the data correctly.

**Comment by sffc** [comment]
OK so regarding the week numbering things.

Wikipedia has a good article on this subject: https://en.wikipedia.org/wiki/Week#Other_week_numbering_systems

There are multiple possible schemes. ISO-8601 is the most common, but there are others. The Financial, Media, and Public Health sectors all have standards that they follow, and they might not be the same.

I think it is very safe to say that in cases when the developer wants to display week numbering, they will also want to opt into a specific week numbering scheme, not just allow the locale to pick. This is how I came up with my "week calendar" suggestion.

This still needs more research and I undo my LGTM on the discussion earlier.

**Comment by sffc** [comment]
Yeah, I'm not sure why I was briefly convinced thismorning that week numbering was a display concern. It absolutely is not! Exclamation point there. It is a data model concern. I think DayOfYearInfo fudges around the problem such that it is not as horrible as it could be, but we really should seek out an actually correct solution.

**Comment by sffc** [comment]
Here's how I envision the use case.

You are a public health agency and wish to use a week calendar for your web site.

You create a `Date<PublicHealthGregorian>`. That object supports a number of `GetField` impls, most importantly `GetField<WeekInfo>`. You are then able to format with a field set requiring that field.

```rust
pub struct WeekInfo {
    year_of_week: YearInfo,
    week_of_year: i8,
    day_of_week_number: u8,
}
```

This struct returns data sufficient to format "3rd day of the 5th week of 2025", using the fields `c`, `w`, and `Y`, none of which are currently supported. This could also be extended to support `W` pending a question I posted to CLDR.

Do we put `PublicHealthGregorian` in `AnyCalendar`? I think so, probably. One catch is that `GetField<WeekInfo>` would be supported for only a subset of `Date<AnyCalendar>`, which I don't think we currently support. However, it may be okay for us to say, we simply don't support locale-dependent selection of week numbering. If you want week numbering, be explicit about it.

**Comment by Manishearth** [comment]
That still feels like a locale to me, `en-US-publichealth` or something. Or a setting thing that you can override with Options. 

I see where you're coming from though.

>  However, it may be okay for us to say, we simply don't support locale-dependent selection of week numbering. If you want week numbering, be explicit about it.

I kind of like this.

**Comment by sffc** [comment]
Without defining a new calendar, we could just model it as a `pub struct Week` living alongside `pub struct Date`. It could be a type defining a specific week in time, not a date.

Not sure what the inner representation would be. It needs the year, but not the month and day. We need to figure this out not only for `Week` but also for `YearMonth`, which we should also probably add at some point.

**Comment by sffc** [comment]
Potential data model:

```rust
pub struct IsoWeek {
    week: i32, // an absolute week number, defining week 0 to be the first week of 1970 or simething
}

pub struct WeekNumberingRule {
    pub weekday_in_first_week: IsoWeekday,
    pub first_day_of_week: IsoWeekday,
    // TODO: Investigate what else we need to fully define the week numbering
}

pub struct Week<A: AsCalendar> {
    week: IsoWeek,
    rule: WeekNumberingRule,
    calendar: A,
}

pub struct WeekInfo {
    pub year_of_week: YearInfo,
    pub week_of_year: i8,
}

impl<A> GetField<WeekInfo> for Week<A> { ... }
```

**Comment by Manishearth** [comment]
> Without defining a new calendar, we could just model it as a `pub struct Week` living alongside `pub struct Date`. It could be a type defining a specific week in time, not a date.

That makes some more sense to me. I'd _really_ love to decouple Week stuff from Date.

**Comment by Manishearth** [comment]
More on that:

To me, a calendar defines:

 - What constitutes a year: When they start, when they end, how they are numbered
 - What constitutes a month
 - What constitutes a day (we don't support calendars that redefine days, but this does exist!)


These are three concepts that stack neatly on top of each other, and can stack on top of the general concept of hour/minute/second time.

Weeks, on the other hand, are not calendar-specific. The concept of a 7-day week has a single origin and has proven incredibly sticky historically, with attempts to unseat it (French Revolutionary Calendar, Soviet Calendar) failing. There are some calendars with other non-month ways of grouping days cyclically (Javanese Calendar), but that's a more general concept, not a "week".

Weeks themselves layer on top of calendars, independently of the calendar: a given day is Monday regardless of the calendar it's in. Our Islamic calendar code literally differentiates epoch dates as "Thursday" and "Friday"!

So weeks are more like `related_iso` for cyclic calendars, an entirely separate timekeeping system whose overlap with the main calendar is useful for disambiguation or redundancy.

Week _numbering_ is a curious concept from the merging of year boundaries (from a specific calendar) and weeks (pan-calendar). It is locale- and calendar- specific, but I don't think it's just a property of the calendar, it's something else. The public health thing to me reads strongly as a preference.

Decoupling week stuff from date stuff would be pretty neat.

**Comment by sffc** [comment]
I agree with @Manishearth's mental model overall.

Regarding week numbering: there are various ways to anchor week numbers. They can be anchored to a year, but this is not the only choice: they are often anchored to a month, and they could in principle be anchored to something else entirely such as religious holidays (ever heard the church saying "Third Sunday of Easter"?) or an astronomical event such as the winter solstice.

So, in that sense, week numbering really is just another calendar system. The periods are the same length, but the way they are anchored and numbered is different.

**Comment by Manishearth** [comment]
> ever heard the church saying "Third Sunday of Easter"?

note: that's because the Christian liturgical year is basically its own somewhat strange calendar system.

![image](https://github.com/user-attachments/assets/1f12b798-7de4-4d39-8f43-92a7d6fbf20d)
But yes, week numbering can be anchored to anything. So can day numbering, but day numbering has some more rigid patterns.

**Comment by sffc** [comment]
Why this is in the 2.0 milestone:

> 2.0 work is to load the data correctly.

and

> I would like to at least come up with a design for this before 2.0.

I am currently working with the CLDR Design WG to nail down what exactly we mean when we talk about week numbering. I can follow up here soon.

**Comment by sffc** [comment]
CLDR Design WG decided that ISO week numbering should be used for all week-of-year numbering.

It therefore does not require locale data, and I will move this issue out of the milestone.

**Comment by iyzana** [comment]
Whether a particular January 1st lies in week-of-year 1 or in week-of-year 53 is locale specific, is it not? Does that mean I won't be able to get that data from icu4x v2 anymore?

**Comment by sffc** [comment]
> Whether a particular January 1st lies in week-of-year 1 or in week-of-year 53 is locale specific, is it not? Does that mean I won't be able to get that data from icu4x v2 anymore?

You can see some of the reasoning in https://unicode-org.atlassian.net/browse/CLDR-18275

I did research for the CLDR Design WG and could not find evidence that users _expect_ a particular week numbering system outside of regions where "first Thursday" week numbering is the norm. In other words, CLDR has had this data for a long time, but I couldn't find evidence of it being used anywhere, and in fact the only evidence I found was from major clients moving away from locale-dependent week numbers.

If you have any evidence that this data is still need it, please share; I'd love to see.

**Comment by iyzana** [comment]
No, I don't have any evidence for that. I'm from Germany where we do use the first Thursday week numbering.

But I'm currently using the CLDR data for rendering a calendar component. This is what it currently looks like for the en-US locale, i.e. with Sunday as the first day of week.

![image(2).png](https://github.com/user-attachments/assets/0fd31b13-5fc9-4982-8ed4-04b371785493)

ISO week numbering always assumes Monday as first day of week right?
So without that data there are no week numbers that are aligned with the weeks I'm displaying.
Would the solution be to hide week numbers for locales where the first day of week is not Monday because it is not culturally relevant?

**Comment by sffc** [comment]
> Would the solution be to hide week numbers for all locales where the first day of week is not Monday because it is never culturally relevant there?

Either that, or display the ISO week numbers anyway. Monday to Thursday will always be correct; only the weekend days will have discrepancies, where week numbers are typically less important. 

If you are able to get data on your user expectations, again, please share, because I was unable to find reliable data for week number expectations of users outside of ISO.

**Comment by sffc** [comment]
Some discussion on week numbering:

- @sffc What do we think of having week numbering only on the ISO calendar, to make it more clear
- @hsivonen CLDR says Gregorian is the locale calendar. Many European locales use ISO week parameters for the Gregorian calendar. @sffc has previously said that ISO week numbering are more clear in en-US to avoid surprising behavior. If you express dates with week numbers, ISO weeks are the only sensible way to do that. So it seems that it is useful to expose ISO weeks on the Gregorian calendar and only expose ISO weeks on the Gregorian calendar.
- @sffc If we have week numbers in the Gregorian calendar, then I agree that we should use ISO week numbers. I haven't heard any evidence yet to the contrary, even though I am open to hearing it.
- @robertbastian Two things here... (a) do we ever want to compute something that isn't the ISO week number? (b) do we want to expose the ISO week number on non-ISO calendars? If this is the only widely used week numbering scheme, it should be available without converting to ISO.
- @sffc What is the behavior on Hijri or Hebrew?
- @robertbastian It should return `IsoWeek` with an ISO week number and ISO year.
- @hsivonen Does your opinion @sffc change if the method is called something like `get_iso_week`?
- @sffc Yes, absolutely. If we had such a method, that is what it should be called. If the value is specific to a calendar system, then the API name should specify that. In order to calculate the ISO week, you have to convert out of the original calendar system anyways. Converting to the ISO calendar is most (80-90%) of the work to convert from a Chinese calendar to get the week number.
- @hsivonen The way I think about it, CLDR says the Gregorian calendar is the preferred calendar except for a few regions. In all of those regions, where week numbers are used, ISO weeks are used. So for those locales, the calendar you have... whether you have the ISO calendar or Gregorian calendar, it's not a relevant answer if you talk about years before 1 CE. And it's useful to have a week numbering function. Overall I think it's useful to have this functionality on Gregorian even if not on other calendars due to the common usage.
- @robertbastian This seems like a question between Temporal and CLDR.
- @nekevss Similar to that, what is the Temporal API vs what is the ICU4X Calendar API? The implicit conversion feels more JavaScript-y than it does Rust.
- @sffc What we currently have in ICU4X is my preferred outcome. We only have one function implemented on `Date<Iso>`. This is new API that I only added a week ago. We only support week numbers on Iso, and I made it on `Date<Iso>` to make that explicit. That is why we are spending meeting time on this. In order to make a generic `week_of_year` that is generic for all calendars is something that would take more design and more work, so I don't want to discuss that right here.
- @robertbastian How about for Japanese?
- @Manishearth It gets weird since eras start in the middle of the ISO year like Reiwa 1
- @manishearth The problem is that `iso_week_of_year` is ambiguous about what the `year` is. It could be ISO year, or regular year, in Gregorian they are exactly the same. In Buddhist they are never the same, and in Japanese they are tantalizingly almost the same except every new era. If we called it `iso_week_of_iso_year` that would be unambiguous but rather weird.
- @sffc If someone proposes evidence of ISO weeks being used in other ISO-like calendars (Buddhist, Japanese, ...), similar functions could be added for them. Needs more research.
- @hsivonen I think there needs to be relevant usage scenarios.

Conclusion:

- ICU4X only calculates ISO week numbers for now
- It is currently exposed on `Date::<Iso>::week_of_year`
- OK to add `Date::<Gregorian>::iso_week_of_year`

LGTM: @sffc @hsivonen @Manishearth @nekevss
